### PR TITLE
fix(server): respect sockPath on transportMode: 'ws' (#2310)

### PR DESCRIPTION
**Pull request by joonas-lahtinen**
- [x] This is a **bugfix**
- [ ] This is a **feature**
- [ ] This is a **code refactor**
- [x] This is a **test update**
- [ ] This is a **docs update**
- [ ] This is a **metadata update**

### For Bugs and Features; did you add new tests?

Added tests to verify that sockPath is respected.

### Motivation / Use-Case

Fixes bug #2310.

### CODE REVIEW

**Comment by evilebottnawi** [comment]
Please accept CLA

File: lib/servers/WebsocketServer.js [hunk 1]
```diff
@@ -3,22 +3,38 @@
...
+      });
+    });
+
     this.wsServer.on('error', (err) => {
       this.server.log.error(err.message);
     });
   }
 
+  shouldUpgrade(req) {
+    const pathname = url.parse(req.url).pathname;
```
Comment (hiroppy) [inline lib/servers/WebsocketServer.js]: memo: we'll use URLSearchParams after major release.

**Comment by joonas-lahtinen** [comment]
Thanks for the quick reviews. I noticed a bug in the tests based on the code coverage difference report, should be fixed now and using the more forward compatible ws.Server.shouldHandle()

File: lib/servers/WebsocketServer.js (hunk 1 above)
Comment (joonas-lahtinen) [reply lib/servers/WebsocketServer.js]: I fixed the code to use WebSocket.Server.shouldHandle()
https://github.com/websockets/ws/blob/master/doc/ws.md#servershouldhandlerequest
Should be more futureproof?

**Comment by joonas-lahtinen** [comment]
This also fixes the bug that whole `webpack-dev-server` process exits due to an unhandled error if the proxy target is not open:

`[HPM] Rewriting path from "/api" to ""
[HPM] GET /api ~> http://localhost:8888
[HPM] Upgrading to WebSocket
events.js:174
      throw er; // Unhandled 'error' event
      ^

Error [ERR_STREAM_DESTROYED]: Cannot call write after a stream was destroyed
`

**Comment by Loonride** [comment]
@joonas-lahtinen I ran your new tests without the changes that you made, and they passed. How are your changes different from the existing behavior? The `ws` docs say for the path option - "Accept only connections matching this path" (https://github.com/websockets/ws/blob/master/doc/ws.md#new-websocketserveroptions-callback), so paths that do not match the one specified are already rejected.

My suggestion is simply to remove the changes you made but keep the tests for us to merge, because the tests are useful for articulating this.

**Comment by joonas-lahtinen** [comment]
Thanks for the patience. After some digging I found the spot to inject testing the different transportMode options with proxy.

**Comment by Loonride** [comment]
LGTM /cc @evilebottnawi @hiroppy

**Comment by hiroppy** [comment]
thanks

**Pull request merged**